A29-xds-tls-security.md

```diff
@@ -0,0 +1,411 @@
+to be used by the plugin to monitor the file paths to refresh the certificates.
+
+#### Use of Bootstrap File by the Certificate Provider Plugin Framework
+
+The xDS control plane specifies a certificate_provider_instance name as described above in
+[CommonTlsContext Processing](#commontlscontext-processing). For this example, let's assume
+it uses `"google_cloud_private_spiffe"` as the `"instance_name"` in
+`tls_certificate_certificate_provider_instance`. gRPC looks up
+`"google_cloud_private_spiffe"` in the `"certificate_providers"` map loaded from the
+bootstrap file and use the JSON object value to extract the plugin name (value of
+`"plugin_name"` as a string) and configuration (value of `"config"` as a JSON object).
+It passes on these values to the `CertificateProviderStore` to retrieve
+(after instantiating if necessary) the associated `CertificateProvider` plugin.
+The caller of the framework registers a watcher with the plugin to receive
+certificate updates as described below.
+
+### Implementing Security in the xDS Flow
+
+Let's see how gRPC enables security in xDS managed clients and servers by putting it all
+together. Note that gRPC uses the existing low level TLS handshaker code/libraries
+(aka security connector or handshaker) after making the certificates and keys available
+to that code. The TLS handshaker functionality (which is language dependent) is
+out of scope of this document.
+
+#### Client Side Flow
+
+The gRPC client xDS flow is described in
+[gRPC Client Architecture](./A27-xds-global-load-balancing.md#grpc-client-architecture).
+The resolver returns a `Cluster` resource for a channel and the CDS for that cluster has the
+[UpstreamTlsContext](https://github.com/envoyproxy/envoy/blob/35702fed462f63a0a237cfbfdf26184272207c11/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L27)
+containing the security configuration to be applied to all the child policies of that cluster.
+When a channel is using `XdsChannelCredentials`, gRPC processes the `UpstreamTlsContext`
+as described [above](#cds-for-client-side-security).
+
+At this stage the CDS policy uses the flow described above in
+[Use of Bootstrap File by the Certificate Provider Plugin Framework](#use-of-bootstrap-file-by-the-certificate-provider-plugin-framework)
+to fetch the requisite `CertificateProvider`s (one or two as the case may be). These `CertificateProvider`s,
+in one form or another, need to be made available to all the sub-channels of the cluster. How this is
+done is language dependent. For example, C-core uses the channel args to pass the individual
+`CertificateProvider`s whereas Java constructs a dynamic `SslContextProvider` that is directly
+usable by the sub-channel's TLS handshaker to build an `SslContext` for the impending handshake.
+
+##### Server Authorization aka Subject-Alt-Name Checks
+
+If [match_subject_alt_names](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L373)
```

Comment:
Don't know how I missed this comment earlier.

> What if it isn't present?  Is no verification of the server performed? 

The server cert trust verification is performed of course. But there is no verification (or rather matching) of the ID presented in the certificate because we don't know what to match it to.

>  ... verify_certificate_hash, verify_certificate_spki....

verify_certificate_hash, verify_certificate_spki are purely cryptographic verification of the peer certificate based on out of band values supplied by the control plane and I don't think is a substitute for either hostname verification in Web PKI or the subject-alt-name check here.

However I should probably add here that we should NACK an update that has these fields populated. Thanks for pointing that out.

Reply:
Done. PTAL